src/front/wgsl/mod.rs

```diff
@@ -379,55 +379,49 @@ impl<'a> ExpressionContext<'a, '_, '_> {
 }
 
 enum Composition {
-    Single(crate::Expression),
-    Multi(crate::VectorSize, Vec<Handle<crate::Expression>>),
+    Single(u32),
+    Multi(crate::VectorSize, [crate::SwizzleComponent; 4]),
 }
 
 impl Composition {
     //TODO: could be `const fn` once MSRV allows
-    fn letter_pos(letter: char) -> u32 {
+    fn letter_component(letter: char) -> Option<crate::SwizzleComponent> {
+        use crate::SwizzleComponent as Sc;
         match letter {
-            'x' | 'r' => 0,
-            'y' | 'g' => 1,
-            'z' | 'b' => 2,
-            'w' | 'a' => 3,
-            _ => !0,
+            'x' | 'r' => Some(Sc::X),
+            'y' | 'g' => Some(Sc::Y),
+            'z' | 'b' => Some(Sc::Z),
+            'w' | 'a' => Some(Sc::W),
+            _ => None,
         }
     }
 
-    fn extract(
-        base: Handle<crate::Expression>,
-        base_size: crate::VectorSize,
-        name: &str,
-        name_span: Range<usize>,
-    ) -> Result<crate::Expression, Error> {
+    fn extract_impl(name: &str, name_span: Range<usize>) -> Result<u32, Error> {
         let ch = name
             .chars()
             .next()
             .ok_or_else(|| Error::BadAccessor(name_span.clone()))?;
-        let index = Self::letter_pos(ch);
-        if index >= base_size as u32 {
-            return Err(Error::BadAccessor(name_span));
```

Comment:
So is the idea that the expression validator now checks that the accessors make sense for the input vector type, so we don't need to check for that here?

Reply:
right, we don't want to implement the same check in multiple places if this can be avoided